allennlp/models/coreference_resolution/coref.py

```diff
@@ -586,40 +602,49 @@ def decode(self, output_dict: Dict[str, torch.Tensor]):
             which are in turn comprised of a list of (start, end) inclusive spans into the
             original document.
         """
-        top_spans = output_dict["top_spans"].data.cpu()
-        predicted_antecedents = output_dict["predicted_antecedents"].data.cpu()
+        batch_top_spans = output_dict["top_spans"].data.cpu()
+        batch_predicted_antecedents = output_dict["predicted_antecedents"].data.cpu()
+        antecedent_indices = output_dict["antecedent_indices"].data.cpu()
         batch_clusters: List[List[List[Tuple[int, int]]]] = []
 
         # Calling zip() on two tensors results in an iterator over their
         # first dimension. This is iterating over instances in the batch.
-        for spans, span_antecedents in zip(top_spans, predicted_antecedents):
-
-            clusters: DefaultDict[int, Set[Tuple[int, int]]] = defaultdict(set)
-            for span, antecedent_index in zip(spans, span_antecedents):
-                if antecedent_index != -1:
-                    # Find the right cluster to update with this span.
-                    # We might have referred to a span which in turn
-                    # refers to some previous span, so here we trace back
-                    # through the list to find the "root" of the cluster,
-                    # and only update that set.
-                    cluster_index_to_update = antecedent_index
-                    # This while loop must halt because antecedents are
-                    # strictly less than the current index, so the cluster
-                    # we are referring to must decrease at each step.
-                    while span_antecedents[cluster_index_to_update] != -1:
-                        cluster_index_to_update = span_antecedents[cluster_index_to_update]
-
-                    # Add the root span to the set, as we might not have seen it before.
-                    root_span_start, root_span_end = spans[cluster_index_to_update]
-                    clusters[cluster_index_to_update].add((root_span_start, root_span_end))
-                    # Now add the span we are currently considering.
-                    span_start, span_end = span
-                    clusters[cluster_index_to_update].add((span_start, span_end))
+        for top_spans, predicted_antecedents in zip(batch_top_spans, batch_predicted_antecedents):
+            spans_to_cluster_ids: Dict[Tuple[int, int], int] = {}
+            clusters: List[List[Tuple[int, int]]] = []
+
+            for i, (span, predicted_antecedent) in enumerate(zip(top_spans, predicted_antecedents)):
+                if predicted_antecedent < 0:
+                    # We don't care about spans which are
+                    # not co-referent with anything.
+                    continue
+
+                # Find the right cluster to update with this span.
+                # To do this, we find the row in ``antecedent_indices``
```

Comment:
I'm having a hard time following what you're saying here, probably because there's no clear description above of what "antecedent_indices" _is_.